File: leet_code/trees/maximum_depth.rb
def max_depth(root)
    return 0 if !root
    answer, depth = 0, 0

    traverse = lambda do |node, depth|
        return depth if !node
        left_max = traverse.call(node.left, depth+1)
        right_max = traverse.call(node.right, depth+1)
        answer = [left_max, right_max].max
    end
    traverse.call(root, depth)
end

File: leet_code/trees/maximum_depth.js
const maxDepth = root => {
    if (!root) return 0;
    let ans = 0,
        depth = 0;

    const traverse = (node, depth) => {
        if (!node) return depth;
        const left = traverse(node.left, depth + 1);
        const right = traverse(node.right, depth + 1);
        const tempMax = Math.max(left, right);
        if (tempMax > ans) ans = tempMax;
    };
    traverse(root, depth);
    return ans;
};
